### PR TITLE
Carlo  trajectory following

### PR DESCRIPTION
**Pull request by CarloWoolsey**
Added odometry as well as path following framework into Drive Train subsystem, created example Ramsete command

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 1]
```diff
@@ -6,30 +6,145 @@
...
-  private SpeedControllerGroup leftMotors = new SpeedControllerGroup(leftFrontMotor, leftBackMotor);
-  private SpeedControllerGroup rightMotors = new SpeedControllerGroup(rightFrontMotor, rightBackMotor);
+  private final SpeedControllerGroup leftMotors = new SpeedControllerGroup(leftFrontMotor, leftBackMotor);
+  private final SpeedControllerGroup rightMotors = new SpeedControllerGroup(rightFrontMotor, rightBackMotor);
+  
+  private final DifferentialDrive differentialDrive = new DifferentialDrive(leftMotors, rightMotors);
 
-  private DifferentialDrive differentialDrive = new DifferentialDrive(leftMotors, rightMotors);
+  private final Encoder leftEncoder = new Encoder(0,1);
+  private final Encoder rightEncoder = new Encoder(2, 3);
```
Comment (Dan-The-Alien) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: Please put ports in the constants class

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 2]
```diff
@@ -6,30 +6,145 @@
...
+  
+//creates odometry class
+  private final DifferentialDriveOdometry m_odometry;
+
+
+ 
+  private final PigeonIMU pigeonIMU = new PigeonIMU(0);
```
Comment (Dan-The-Alien) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: Again, ports in constants.java

**Review by Dan-The-Alien — CHANGES_REQUESTED** [review]
Hey Carlo, nice work! Hopefully it will make our robot autonomous :) If you could just please place the encoder and IMU ports in the Constants class, then I can accept the pull request. Once you do that, just commit it as you would normally, and GitHub will probably re-open the pull request. Thanks

File: src/main/java/frc/robot/subsystems/DriveTrain.java (hunk 1 above)
Comment (CarloWoolsey) [reply src/main/java/frc/robot/subsystems/DriveTrain.java]: I created constants for the pigeonIMU as well as encoder ports.

**Pull request merged**